checker/cost.go

```diff
@@ -0,0 +1,536 @@
+	case *exprpb.Expr_ListExpr:
+		cost = c.costCreateList(e)
+	case *exprpb.Expr_StructExpr:
+		cost = c.costCreateStruct(e)
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
+	// Check to see if the identifier is declared.
+	if ident := c.env.LookupIdent(identExpr.GetName()); ident != nil {
+		c.setType(e, ident.GetIdent().Type)
+		// Overwrite the identifier with its fully qualified name.
+		identExpr.Name = ident.GetName()
+		return identCost
+	}
+
+	// undeclared reference
+	c.setType(e, decls.Error)
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+	// Before traversing down the tree, try to interpret as qualified name.
+	qname, found := containers.ToQualifiedName(e)
+	if found {
+		ident := c.env.LookupIdent(qname)
+		if ident != nil {
+			if sel.TestOnly {
+				// expression does not select a field
+				c.setType(e, decls.Bool)
+				return sum
+			}
+			// Rewrite the node to be a variable reference to the resolved fully-qualified
+			// variable name.
+			c.setType(e, ident.GetIdent().Type)
+			identName := ident.GetName()
+			e.ExprKind = &exprpb.Expr_IdentExpr{
+				IdentExpr: &exprpb.Expr_Ident{
+					Name: identName,
+				},
+			}
+			return sum.Add(selectCost)
+		}
+	}
+
+	// Interpret as field selection, first traversing down the operand.
+	sum = sum.Add(c.cost(sel.Operand))
+	targetType := c.getType(sel.Operand)
+	// Assume error type by default as most types do not support field selection.
+	resultType := decls.Error
+	switch kindOf(targetType) {
+	case kindMap:
+		// Maps yield their value type as the selection result type.
+		mapType := targetType.GetMapType()
+		resultType = mapType.ValueType
+		sum.Add(selectCost)
+	case kindObject:
+		// Objects yield their field type declaration as the selection result type, but only if
+		// the field is defined.
+		messageType := targetType
+		if fieldType, found := c.lookupFieldType(
+			messageType.GetMessageType(),
+			sel.Field); found {
+			resultType = fieldType.Type
+		}
+		sum.Add(selectCost)
+	case kindTypeParam:
+		// Set the operand type to DYN to prevent assignment to a potentionally inco
+		//rrect type
+		// at a later point in type-checking. The isAssignable call will update the type
+		// substitutions for the type param under the covers.
+		c.isAssignable(decls.Dyn, targetType)
+		// Also, set the result type to DYN.
+		resultType = decls.Dyn
+		sum.Add(selectCost)
+	default:
+		// Dynamic / error values are treated as DYN type. Errors are handled this way as well
+		// in order to allow forward progress on the cost.
+		if isDynOrError(targetType) {
+			resultType = decls.Dyn
+		} else {
+			// does not support field selection
+		}
+	}
+	if sel.TestOnly {
+		resultType = decls.Bool
+	}
+	c.setType(e, resultType)
```

Comment:
I would just make the `types` immutable (skip setting the type or computing the result type) since the information should already be filled in by the type-checker.

Reply:
I was _almost_ able to make this work. I have one remaining setType that in costIdent that ensures that the type of identifiers is the exact same type reference that was provided by the typeProvider. I haven't isolated why this is needed yet. It might be better to look at this one together offline as it's a bit tough to explain.